File: Services/pkg/role/cache.go
package role

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/ChangerzaryX1602/SkillSync/pkg/models"
	"github.com/ChangerzaryX1602/SkillSync/pkg/utils"
	"github.com/gofiber/storage/redis"
)

const (
	cacheTTL     = 15 * time.Minute
	cacheTTLList = 1 * time.Minute
)

type roleCache struct {
	store *redis.Storage
}

func newRoleCache(store *redis.Storage) *roleCache {
	return &roleCache{store: store}
}

func (c *roleCache) Get(id uint) (*models.Role, bool) {
	if c.store == nil {
		return nil, false
	}
	key := fmt.Sprintf("%s:%d", models.PkgRoleGetRole, id)

	bytes, err := c.store.Get(key)
	if err != nil || len(bytes) == 0 {
		return nil, false
	}

	var role models.Role
	if err := json.Unmarshal(bytes, &role); err != nil {
		return nil, false
	}
	return &role, true
}

func (c *roleCache) Set(role *models.Role) {
	if c.store == nil || role == nil {
		return
	}

	go func(r models.Role) {
		key := fmt.Sprintf("%s:%d", models.PkgRoleGetRole, r.ID)
		if bytes, err := json.Marshal(r); err == nil {
			_ = c.store.Set(key, bytes, utils.RandomJitter(cacheTTL))
		}
	}(*role)
}

func (c *roleCache) GetByName(name string) (*models.Role, bool) {
	if c.store == nil {
		return nil, false
	}
	key := fmt.Sprintf("%s:%s", models.PkgRoleGetRoleByName, name)

	bytes, err := c.store.Get(key)
	if err != nil || len(bytes) == 0 {
		return nil, false
	}

	var role models.Role
	if err := json.Unmarshal(bytes, &role); err != nil {
		return nil, false
	}
	return &role, true
}

func (c *roleCache) SetByName(name string, role *models.Role) {
	if c.store == nil || role == nil {
		return
	}

	go func(r models.Role) {
		key := fmt.Sprintf("%s:%s", models.PkgRoleGetRoleByName, name)
		if bytes, err := json.Marshal(r); err == nil {
			_ = c.store.Set(key, bytes, utils.RandomJitter(cacheTTL))
		}
	}(*role)
}

func (c *roleCache) Invalidate(id uint, name string) {
	if c.store == nil {
		return
	}
	go func() {
		key := fmt.Sprintf("%s:%d", models.PkgRoleGetRole, id)
		_ = c.store.Delete(key)
	}()
	go func() {
		key := fmt.Sprintf("%s:%s", models.PkgRoleGetRoleByName, name)
		_ = c.store.Delete(key)
	}()
	c.InvalidateAllLists()
}
func (c *roleCache) InvalidateAllLists() {
	if c.store == nil {
		return
	}
	pattern := fmt.Sprintf("%s:*", models.PkgRoleGetRoles)
	conn := c.store.Conn()
	ctx := context.Background()
	keys, err := conn.Keys(ctx, pattern).Result()
	if err != nil {
		return
	}
	if len(keys) == 0 {
		return
	}
	const numWorkers = 100
	keyChan := make(chan string, len(keys))
	var wg sync.WaitGroup
	workerCount := numWorkers
	if len(keys) < numWorkers {
		workerCount = len(keys)
	}
	for i := 0; i < workerCount; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for key := range keyChan {
				_ = c.store.Delete(key)
			}
		}()
	}
	for _, key := range keys {
		keyChan <- key
	}
	close(keyChan)
	wg.Wait()
}
func (c *roleCache) GetList(pagination models.Pagination, search models.Search) ([]models.Role, bool) {
	if c.store == nil {
		return nil, false
	}
	key := fmt.Sprintf("%s:%s:%s", models.PkgRoleGetRoles, pagination.GetPaginationString(), search.GetSearchString())

	bytes, err := c.store.Get(key)
	if err != nil || len(bytes) == 0 {
		return nil, false
	}

	var roles []models.Role
	if err := json.Unmarshal(bytes, &roles); err != nil {
		return nil, false
	}
	return roles, true
}

func (c *roleCache) SetList(pagination models.Pagination, search models.Search, roles []models.Role) {
	if c.store == nil {
		return
	}
	go func(rs []models.Role) {
		key := fmt.Sprintf("%s:%s:%s", models.PkgRoleGetRoles, pagination.GetPaginationString(), search.GetSearchString())
		if bytes, err := json.Marshal(rs); err == nil {
			_ = c.store.Set(key, bytes, utils.RandomJitter(cacheTTLList))
		}
	}(roles)
}

File: hono-clean-architecture/pkg/role/cache.ts
import Redis from "ioredis";
import { Role } from "../models";
import { randomJitter } from "../utils";

const CACHE_TTL = 15 * 60 * 1000;

export class RoleCache {
  private store: Redis | null;

  constructor(store: Redis | null) {
    this.store = store;
  }

  async get(id: number): Promise<Role | null> {
    if (!this.store) return null;

    const key = `pkg:role:get:${id}`;
    const data = await this.store.get(key);

    if (!data) return null;

    try {
      return JSON.parse(data) as Role;
    } catch {
      return null;
    }
  }

  async set(role: Role): Promise<void> {
    if (!this.store || !role) return;

    const key = `pkg:role:get:${role.id}`;
    const ttl = randomJitter(CACHE_TTL);

    await this.store.set(key, JSON.stringify(role), "PX", ttl);
  }

  async getByName(name: string): Promise<Role | null> {
    if (!this.store) return null;

    const key = `pkg:role:name:${name}`;
    const data = await this.store.get(key);

    if (!data) return null;

    try {
      return JSON.parse(data) as Role;
    } catch {
      return null;
    }
  }

  async setByName(name: string, role: Role): Promise<void> {
    if (!this.store || !role) return;

    const key = `pkg:role:name:${name}`;
    const ttl = randomJitter(CACHE_TTL);

    await this.store.set(key, JSON.stringify(role), "PX", ttl);
  }

  async invalidate(id: number, name?: string): Promise<void> {
    if (!this.store) return;

    const keys: string[] = [`pkg:role:get:${id}`];
    if (name) {
      keys.push(`pkg:role:name:${name}`);
    }

    await this.store.del(...keys);
  }

  async invalidateAllLists(): Promise<void> {
    if (!this.store) return;

    const pattern = `pkg:role:list:*`;
    const keys = await this.store.keys(pattern);

    if (keys.length > 0) {
      await this.store.del(...keys);
    }
  }
}

export function newRoleCache(store: Redis | null): RoleCache {
  return new RoleCache(store);
}
